Biteable/deploy-buildkite-plugin
# [ENG-132] Git checking

## High Level Description

Git checking for builds/deploys.

## Detailed Description

ENG-132

Verify that git commit being deployed is the desired one
    
Builds using this plugin will fail if the checkout commit does not match what's expected.
    
This was adapted from another project (biteable-micro-frontends).
    
Some of the conditions should not be possible on the build server (e.g. "uncommitted changes") but are retained for completeness.
    
The check for branch equality is also removed - that fails on build server because of the way buildkite checkouts a single commit rather than a branch - commit comparison should be enough (even if it's theoretically possible that the branches differ).
    
We have to do a full `git fetch origin` here too, so that we can compare with stuff on origin (else, we don't have it locally).

## Test Instructions

Add this plugin to another project pipeline with config something like the below:

```yml
      - Biteable/deploy#git-checking:
          deploy_branch: <made up branch> # build will fail
```

or to verify the happy path works:

```yml
      - Biteable/deploy#git-checking:
          deploy_branch: <your actual branch> # build will pass
```

Depending on the project config, you might need to remove the branch restriction and change the deploy command to a dummy command, e.g.:

```yml
  - label: "Deploy"
    # command: "bin/deploy.sh"
    command: "echo testing"
    # branches: "master"
```
